prow/bugzilla/client.go

```diff
@@ -220,31 +275,106 @@ func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
 func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+	bugCache := NewBugDetailsCache()
+	return getAllClones(c, bug, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+func getAllClones(c Client, bug *Bug, bugCache *BugDetailsCache) (clones []*Bug, err error) {
+
+	clones = []*Bug{}
+	err = getAllLinkedBugs(c, bug.ID, bugCache, nil)
 	if err != nil {
 		return nil, err
 	}
-	clones, err := getRecursiveClones(c, root)
-	if err != nil {
-		return nil, err
+
+	for _, node := range bugCache.List() {
+		if node.Summary == bug.Summary {
+			clones = append(clones, node)
+		}
 	}
-	// Getting rid of the bug whose clones we are searching for
-	// we could optimize this (maybe?) if the list turns out to be too long
-	indexOfOriginal := -1
-	for index, clone := range clones {
-		if clone.ID == bug.ID {
-			// Check if there is a more elegant way to do this
-			indexOfOriginal = index
+	return clones, nil
+}
+
+// Parallel implementation for getAllClones - spawns threads to go up and down the tree
+// Also parallelizes the getBug calls if bug has multiple bugs in DependsOn/Blocks
+func getAllLinkedBugs(c Client, bugID int, bugCache *BugDetailsCache, errGroup *errgroup.Group) error {
+	var shouldWait bool
+	if errGroup == nil {
+		shouldWait = true
+		errGroup = new(errgroup.Group)
+	}
+	var err error
+	bug, _, cacheHasBug := bugCache.Get(bugID)
+	if !cacheHasBug {
+		bug, err = c.GetBug(bugID)
+		if err != nil {
+			return err
 		}
 	}
-	if indexOfOriginal == -1 {
-		return nil, fmt.Errorf("Original bug not found in list of clones. Error in logic for getRecursiveClones/getRootForClone")
+	errGroup.Go(func() error {
+		return traverseUp(c, bug, bugCache, errGroup)
+	})
+	errGroup.Go(func() error {
+		return traverseDown(c, bug, bugCache, errGroup)
+	})
+
+	if shouldWait {
+		return errGroup.Wait()
 	}
-	clones = append(clones[:indexOfOriginal], clones[indexOfOriginal+1:]...)
-	return clones, nil
+	return nil
+}
+
+func traverseUp(c Client, bug *Bug, bugCache *BugDetailsCache, errGroup *errgroup.Group) error {
+	for _, dependsOnID := range bug.DependsOn {
+		dependsOnID := dependsOnID
+		errGroup.Go(func() error {
+			parent, isVisited, alreadyFetched := bugCache.Get(dependsOnID)
```

Comment:
Why do we need `isVisisted`? Either its in the cache then we fetched it or its not, then we didn't fetch it

Reply:
If in case we get the cache prepopulated with data(which would be the case if we do a prefetch, also the case if we do a GetBug for just the main bug), then isVisited is used to mark whether or not the the getLinkedBugs needs to be called on it.
ie to identify the case where the bug is fetched and we need to still call getLinkedBugs

Reply:
Sorry I don't understand. There is no way to pre-populate the cache. Lets keep this scoped to just the traversal optimization. We always call `getLinkedBugs`. We should use the `alreadyFetched` check for short-circuiting where you are currently using the `isVisited`.